catalyst-toolbox/src/rewards/proposers/mod.rs

```diff
@@ -0,0 +1,288 @@
+
+    let pass_total_threshold = total as f64 >= total_stake_threshold;
+    let pass_relative_threshold = (yes as f64 / no as f64) >= threshold;
+    let success = pass_total_threshold && pass_relative_threshold;
+
+    debug!("total_stake_threshold: {total_stake_threshold}, threshold: {threshold}");
+    debug!("success: {success}, total_threshold: {pass_total_threshold}, relative_threshold: {pass_relative_threshold}");
+
+    Ok((diff, success))
+}
+
+/// returns (yes, no)
+fn extract_yes_no_votes(proposal: &Proposal, voteplan: &VoteProposalStatus) -> Result<(u64, u64)> {
+    let yes_index = proposal
+        .chain_vote_options
+        .0
+        .get("yes")
+        .ok_or(eyre!("missing `yes` field"))?;
+    let no_index = proposal
+        .chain_vote_options
+        .0
+        .get("no")
+        .ok_or(eyre!("missing `no` field"))?;
+
+    let tally = match &voteplan.tally {
+        Tally::Public { result } => result,
+        Tally::Private { .. } => bail!("private vote tally"),
+    };
+
+    let yes_result = tally.results()[*yes_index as usize];
+    let no_result = tally.results()[*no_index as usize];
+
+    Ok((yes_result, no_result))
+}
+
+fn filter_data_by_challenge(
+    challenge_id: i32,
+    proposals: &HashMap<Hash, Proposal>,
+    voteplans: &HashMap<Hash, VoteProposalStatus>,
+) -> (HashMap<Hash, Proposal>, HashMap<Hash, VoteProposalStatus>) {
+    let proposals: HashMap<_, _> = proposals
+        .iter()
+        .filter(|(_, prop)| prop.challenge_id == challenge_id)
+        .map(|(k, v)| (*k, v.clone()))
+        .collect();
+
+    let voteplans = voteplans
+        .iter()
+        .filter(|(_, plan)| proposals.contains_key(&plan.proposal_id))
+        .map(|(k, v)| (*k, v.clone()))
+        .collect();
+
+    (proposals, voteplans)
+}
+
+fn sanity_check_data(
+    proposals: &HashMap<Hash, Proposal>,
+    voteplans: &HashMap<Hash, VoteProposalStatus>,
+) -> Result<()> {
+    let proposals_set: HashSet<_> = proposals.keys().copied().collect();
+    let voteplan_proposals_set: HashSet<_> = voteplans.keys().copied().collect();
+
+    if proposals_set != voteplan_proposals_set {
+        let diff = proposals_set
+            .symmetric_difference(&voteplan_proposals_set)
+            .join(", ");
+        bail!("proposal id set inconsistency. Bad ids: {}", diff);
+    }
+
+    Ok(())
+}
+
+fn filter_excluded_proposals(
+    proposals: &HashMap<Hash, Proposal>,
+    excluded_proposals: &HashSet<String>,
+) -> HashMap<Hash, Proposal> {
+    let predicate = |prop: &Proposal| {
+        let chain_proposal_id = String::from_utf8(prop.chain_proposal_id.clone()).unwrap();
+
+        !excluded_proposals.contains(&prop.proposal_id)
+            && !excluded_proposals.contains(&chain_proposal_id)
+    };
+
+    proposals
+        .iter()
+        .filter(|(_, prop)| predicate(prop))
+        .map(|(id, prop)| (*id, prop.clone()))
+        .collect()
+}
+
+fn calculate_total_stake_from_block0(
```

Comment:
I think we should use `VoterHIR` instead of block0 like we do with other tools

Reply:
I'm not sure what that would look like, is there somewhere I can look for an example of this? Doing a grep for `VoterHir` didn't help much :/

Reply:
Instead of commitee keys + block0 accept as input a `Vec<VoterHIR>`, which is essentially the output from our snapshot tool. Then it should be something like
```
hirs.iter().map(|v| v.voting_power).sum()
```
because we don't punt commitee keys in the snapshot (or rather, all voters in the snapshot should be eligible to vote / earn rewards)

Reply:
@cameron1024 are committee keys in the current snapshot? is this still relevant?